src/controllers/hid/hidioglobaloutputreportfifo.cpp

```diff
@@ -0,0 +1,131 @@
+#include "controllers/hid/hidioglobaloutputreportfifo.h"
+
+#include <hidapi.h>
+
+#include "controllers/defs_controllers.h"
+#include "controllers/hid/legacyhidcontrollermappingfilehandler.h"
+#include "util/compatibility/qbytearray.h"
+#include "util/string.h"
+#include "util/time.h"
+#include "util/trace.h"
+
+namespace {
+constexpr int kReportIdSize = 1;
+constexpr int kMaxHidErrorMessageSize = 512;
+} // namespace
+
+HidIoGlobalOutputReportFifo::HidIoGlobalOutputReportFifo()
+        : m_indexOfLastSentReport(0),
+          m_indexOfLastCachedReport(0) {
+}
+
+void HidIoGlobalOutputReportFifo::addReportDatasetToFifo(const quint8 reportId,
+        const QByteArray& data,
+        const mixxx::hid::DeviceInfo& deviceInfo,
+        const RuntimeLoggingCategory& logOutput) {
+    auto cacheLock = lockMutex(&m_fifoMutex);
+
+    unsigned int indexOfReportToCache;
+
+    if (m_indexOfLastCachedReport + 1 < kSizeOfFifoInReports) {
+        indexOfReportToCache = m_indexOfLastCachedReport + 1;
+    } else {
+        indexOfReportToCache = 0;
+    }
+
+    // Handle the case, that the FIFO is full - which is an error case
+    if (m_indexOfLastSentReport == indexOfReportToCache) {
```

Comment:
Overwriting old, unsent data with new incoming data would probably be the more approriate strategy. A lock-free FIFO wouldn't allow this, but since you are locking here anyway why not use this alternative overflow strategy?

Reply:
This is a very good point and I'm unsure what's the best strategy to handle an overflow:

1. Overwriting the last report ensures, that vital initialization reports at startup will be send if the overflow happens at startup
2. Overwriting the oldest report would propably better recover, when the overflow happens during the session
   This assumens that new report data supersede previous report data, which is the case for class compliant report usage, but the FIFO mode is especially for devices that use reports in a non-class compliant way.